### PR TITLE
Integrate graph objects into hdbscan

### PR DESCRIPTION
**Pull request by JanRhoKa**
Adapt the HDBSCAN function for graphs. 

Create a new metric called `"graph"`, which run HDBSCAN on graphs using an adjacency matrix of the graph, in `scipy.sparse.csr_matrix` format. 

Modified the `_hdbscan_sparse_distance_matrix` function for this, as no graph has to be calculated from the data. Additionally an example of the function is provided under `examples\plot_hdbscan_graph.py` which shows the communities calculated by HDBSCAN, the communities calculated using the precomputed metric and the time save by integrating the graph directly into the function.

### CODE REVIEW

File: hdbscan/hdbscan_.py [hunk 1]
```diff
@@ -162,42 +180,62 @@ def _hdbscan_sparse_distance_matrix(
...
+    if metric != "graph":
+        # Check for connected component on X
+        if csgraph.connected_components(X, directed=False, return_labels=False) > 1:
+            raise ValueError(
+                "Sparse distance matrix has multiple connected "
+                "components!\nThat is, there exist groups of points "
+                "that are completely disjoint -- there are no distance "
+                "relations connecting them\n"
+                "Run hdbscan on each component."
+            )
```
Comment (ivirshup) [inline hdbscan/hdbscan_.py:194]: I would also change make the logic changes a bit lighter here. Something like:

<details>
<summary> alternate changes </summary>

```python
def _hdbscan_sparse_distance_matrix(
    X,
    min_samples=5,
    alpha=1.0,
    metric="minkowski",
    p=2,
    leaf_size=40,
    gen_min_span_tree=False,
    **kwargs
):
    assert issparse(X)

    # if the metric is not graph, compute mutual_reachability of distance matrix
    if metric == "graph":
        mutual_reachability_ = X.tocsr()

    else:
        # Check for connected component on X
        if csgraph.connected_components(X, directed=False, return_labels=False) > 1:
            raise ValueError(
                "Sparse distance matrix has multiple connected "
                "components!\nThat is, there exist groups of points "
                "that are completely disjoint -- there are no distance "
                "relations connecting them\n"
                "Run hdbscan on each component."
            )

        lil_matrix = X.tolil()

        # Compute sparse mutual reachability graph
        # if max_dist > 0, max distance to use when the reachability is infinite
        max_dist = kwargs.get("max_dist", 0.0)
        mutual_reachability_ = sparse_mutual_reachability(
            lil_matrix, min_points=min_samples, max_dist=max_dist, alpha=alpha
        )

    # Check connected component on mutual reachability
    # If more than one component, it means that even if the distance matrix X
    # has one component, there exists with less than `min_samples` neighbors
    if (
            csgraph.connected_components(
                mutual_reachability_, directed=False, return_labels=False
            )
            > 1
    ):
        raise ValueError(
            (
                "There exists points with less than %s neighbors. "
                "Ensure your distance matrix has non zeros values for "
                "at least `min_sample`=%s neighbors for each points (i.e. K-nn graph), "
                "or specify a `max_dist` to use when distances are missing."
            )
            % (min_samples, min_samples)
        )

    # Compute the minimum spanning tree for the sparse graph
    sparse_min_spanning_tree = csgraph.minimum_spanning_tree(mutual_reachability_)

    # Convert the graph to scipy cluster array format
    nonzeros = sparse_min_spanning_tree.nonzero()
    nonzero_vals = sparse_min_spanning_tree[nonzeros]
    min_spanning_tree = np.vstack(nonzeros + (nonzero_vals,)).T

    # Sort edges of the min_spanning_tree by weight
    min_spanning_tree = min_spanning_tree[np.argsort(min_spanning_tree.T[2]), :][0]

    # Convert edge list into standard hierarchical clustering format
    single_linkage_tree = label(min_spanning_tree)

    if gen_min_span_tree:
        return single_linkage_tree, min_spanning_tree
    else:
        return single_linkage_tree, None
```

</details>

I think this implementation makes it more clear that the difference is just computing mutual_reachability or not.

It would probably make sense to update the value error messages so they aren't specific to distance matrices though.

File: hdbscan/tests/test_hdbscan.py [hunk 2]
```diff
@@ -646,6 +643,33 @@ def test_hdbscan_is_sklearn_estimator():
...
+
+    # create a graph from the distance matrix and transform the graph to a csr adjacency matrix
+    graph = nx.from_numpy_matrix(D.toarray())
+    adjacency_matrix = nx.adjacency_matrix(graph)
+
+    # create cluster labels using the graph metric
+    clusterer = HDBSCAN(metric="graph").fit(adjacency_matrix)
+    labels_hdbscan_graph = clusterer.labels_
+
+    assert sklearn.metrics.accuracy_score(labels_distance_matrix, labels_hdbscan_graph) == 1
```
Comment (ivirshup) [inline hdbscan/tests/test_hdbscan.py:669]: I'm not sure this makes sense to me. I think something like:

```suggestion
def test_hdbscan_graph():
    # create a distance matrix, see test_hdbscan_distance_matrix
    D = distance.squareform(distance.pdist(X))
    D /= np.max(D)

    threshold = stats.scoreatpercentile(D.flatten(), 50)

    D[D >= threshold] = 0.0
    D = sparse.csr_matrix(D)
    D.eliminate_zeros()

    # create cluster labels using precomputed metric
    dist_clusterer = HDBSCAN(metric="precomputed").fit(D)
    labels_distance_matrix = dist_clusterer.labels_

    # create a graph from the distance matrix and transform the graph to a csr adjacency matrix
    from hdbscan._hdbscan_reachability import sparse_mutual_reachability

    graph = sparse_mutual_rechablility(D.tolil())

    # create cluster labels using the graph metric
    graph_clusterer = HDBSCAN(metric="graph").fit(graph)
    labels_hdbscan_graph = graph_clusterer.labels_

    assert sklearn.metrics.accuracy_score(labels_distance_matrix, labels_hdbscan_graph) == 1
```

May fit better, and removes the networkx dependency. You could probably also check more things about the output.

More on this in the main review comment

**Review by ivirshup — COMMENTED** [review]
Glad to see this PR opened @JanRhoKa!

I've made a few comments about code organization.

My main thinking here is that the whole idea of this PR is to allow users to weight the graphs by means other than mutual reachability of a distance matrix. I think this could be done with smaller changes to the code in a few places.

For example, the changes I suggested to the test is more "If a user passed mutual reachability weights, the result should be equivalent to passing distances and letting hdbscan compute mutual reachability weights." Right now it looks like you're testing that the result of using mutual reachability weights vs. just the distances directly is minimal.

----------

Minor other point: I would see if you could revert the unrelated formatting changes.

**Comment by lmcinnes** [comment]
Thanks for the updates. This looks like it is starting to come together. The test failures are due to importing networkx in the test module; but it doesn't seem to be actually used there. Perhaps you can remove the import.

It might also be useful (to help ensure the feature sees usage) to include some documentation for this. Perhaps a short tutorial on clustering a graph with hdbscan? Perhaps based on the example you already have?
